src/Microsoft.AspNetCore.WebUtilities/FileBufferingReadStream.cs

```diff
@@ -272,10 +288,27 @@ public override int EndRead(IAsyncResult asyncResult)
             if (_inMemory && _buffer.Length + read > _memoryThreshold)
             {
                 _inMemory = false;
+                var oldBuffer = _buffer;
                 _buffer = CreateTempFile();
-                await _buffer.WriteAsync(_rentedBuffer, 0, (int)_buffer.Length, cancellationToken);
-                _bytePool.Return(_rentedBuffer);
-                _rentedBuffer = null;
+                if (_rentedBuffer == null)
+                {
+                    oldBuffer.Position = 0;
```

Comment:
It breaks my brain a little bit to look at this code and see 'buffer' used for both `Stream` and arrays